redis/db/redisbp/monitored_client.go

```diff
@@ -72,6 +83,75 @@ func (cc *ClusterClient) Wait(ctx context.Context, args WaitArgs) (replicas int6
 	return
 }
 
+var (
+	gaugeLabels = []string{
+		nameLabel,
+	}
+
+	totalConnectionsDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(promNamespace, subsystemPool, "connections"),
+		"Number of total connections in this redisbp pool",
+		gaugeLabels,
+		nil,
+	)
+
+	idleConnectionsDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(promNamespace, subsystemPool, "idle_connections"),
+		"Number of idle connections in this redisbp pool",
+		gaugeLabels,
+		nil,
+	)
+
+	staleConnectionsDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(promNamespace, subsystemPool, "stale_connections"),
+		"Number of stale connections in this redisbp pool",
+		gaugeLabels,
+		nil,
+	)
+)
+
+// Exporter provides an interface for Prometheus merics.
+type Exporter struct {
+	client PoolStatser
+	name   string
+}
+
+func NewExporter(client PoolStatser, name string) *Exporter {
+	return &Exporter{
+		client: client,
+		name:   name,
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	// All metrics are described dynamically.
+}
+
+// Collect implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	stats := e.client.PoolStats()
+
+	ch <- prometheus.MustNewConstMetric(
```

Comment:
is there a reason we have to use the `Must` version, instead of using `promauto` to declare global variables? I really don't like the idea of using `Must*` outside of `init` or global variable declarations.

Reply:
This is the standard for metrics collection when the values you're gathering are external to your application.

Reply:
https://prometheus.io/docs/instrumenting/writing_exporters/#collectors

Reply:
ok looking at https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#NewConstMetric, the only way it would cause panic is label mismatch, which I guess is fine.

but in that case, we should add a test to call this `Collect` function, so that if we actually have a label mismatch in the code, the test would panic (and thus fail).

Reply:
+1 to making this fail at unit test time